drivers/soundwire/slave.c

```diff
@@ -53,6 +53,16 @@ int sdw_slave_add(struct sdw_bus *bus,
 	slave->dev.of_node = of_node_get(to_of_node(fwnode));
 	slave->dev.type = &sdw_slave_type;
 	slave->dev.groups = sdw_slave_status_attr_groups;
+
+	ret = device_register(&slave->dev);
+	if (ret) {
+		dev_err(bus->dev, "Failed to register slave->dev: ret %d\n", ret);
+
+		/* On err, free the slave device. */
+		kfree(slave);
+		return ret;
+	}
+
```

Comment:
this is completely wrong @keyonjie 

First, you didn't check the documentation that explicitly says NOT to use kfree.

https://elixir.bootlin.com/linux/latest/source/drivers/base/core.c#L3175

The put_device() already takes care of freeing the slave, by way of the release function.

Also you need to have the slave structure initialized first, since registering the device will cause the slave driver to probe, and with your code you will probe with invalid information.

If you want to optimize the link list management and only add the slave to the list after successful registration, a tentative fix should be to move the lines 68..70 after line 86, and remove the lines 80..82

````diff
diff --git a/drivers/soundwire/slave.c b/drivers/soundwire/slave.c
index 112b21967c7a..0c92db2e1ddc 100644
--- a/drivers/soundwire/slave.c
+++ b/drivers/soundwire/slave.c
@@ -65,9 +65,6 @@ int sdw_slave_add(struct sdw_bus *bus,
        for (i = 0; i < SDW_MAX_PORTS; i++)
                init_completion(&slave->port_ready[i]);
 
-       mutex_lock(&bus->bus_lock);
-       list_add_tail(&slave->node, &bus->slaves);
-       mutex_unlock(&bus->bus_lock);
 
        ret = device_register(&slave->dev);
        if (ret) {
@@ -77,13 +74,15 @@ int sdw_slave_add(struct sdw_bus *bus,
                 * On err, don't free but drop ref as this will be freed
                 * when release method is invoked.
                 */
-               mutex_lock(&bus->bus_lock);
-               list_del(&slave->node);
-               mutex_unlock(&bus->bus_lock);
                put_device(&slave->dev);
 
                return ret;
        }
+
+       mutex_lock(&bus->bus_lock);
+       list_add_tail(&slave->node, &bus->slaves);
+       mutex_unlock(&bus->bus_lock);
+
        sdw_slave_debugfs_init(slave);
 
        return ret;
````

[sdw.diff.txt](https://github.com/thesofproject/linux/files/6163615/sdw.diff.txt)

Reply:
Thanks for help @plbossart, really good guiding.